Avoid installing tests to site-packages

jdufresne opened the pull request:
For library end users, there is no need to install tests alongside the package itself. This keeps the tests available for development without adding extra packages to user's site-packages directory. Reduces the size of the installed package. Avoids accidental execution of test code by an installed package.

---

Several early commits were done to make the review easier and illustrate the progression of the change. Listed below:

---

Fix use of `async` in test_cursor.py
    
`async` will be a keyword starting with Python 3.7. On Python 3.6, use of `async` causes a deprecation warning. Use the alias `async_` instead.

---

Fix definition of `BUILD_DIR` in `Makefile`
    
The Python's platform should be included in the build dir. Was previously missing resulting in the incorrect path.
    
Fixing the build dir allows testing psycopg2 without first calling `python setup.py install`.
   
Exposed a missing use of the deprecated "async" keyword.
    
For additional information on how to calculate the build dir, see the following SO post:
    
https://stackoverflow.com/a/14369968

---

Avoid installing for tests use the psycopg2 in the build directory
    
Reuses `Makefile` target. Creates a consistent test environment between Travis and development.

Work towards moving the tests outside of the installed package.

---

Use `print()` function instead of print statement throughout project
    
Forward compatible with newer Pythons.

---

Use relative imports throughout tests
    
The tests relied on Python2 relative import semantics. Python3 changed import semantics to always search `sys.path` by default. To import using a relative path it must have a leading dot.
    
Forward compatible with newer Pythons.
    
Works towards the goal of moving tests outside of the installed package.
    
For more information, see PEP-328:
    
https://www.python.org/dev/peps/pep-0328/

---

Skip `register_hstore(..., unicode=True)` tests on Python 3
    
Per the functions documentation, this argument is not supported on Python 3. Skip it during tests.
    
> :param unicode: if `!True`, keys and values returned from the database will be `!unicode` instead of `!str`. The option is not available on Python 3
    
Work towards moving tests outside of the installed package.

---

All feedback welcome. Thanks!

dvarrazzo commented on tests/test_types_extras.py:
```diff
@@ -813,6 +816,30 @@ def _create_type(self, name, fields):
+
+    return skip_if_no_json_module_
+
+
```
Ah, it's already rebased. Then you may want to cherry-pick d0c7126aa88e8347d1add7e7e0f61974717e58d3 and fix-up this change

dvarrazzo commented on .travis.yml:
```diff
@@ -11,7 +11,7 @@ python:
   - 3.4
 
 install:
-  - python setup.py install
```
Yep, actually "I don't mind" is not enough. The tests on travis and appveyor *must* test what is installed, not what was checked out.

jdufresne commented:
> The side effect of these changes is that the test code is no more written in py2 and converted by 2to3 (which was the way to work at the time when the code was ported to py3, I understand it is no more fashionable...). It is a departure from the way the Python code of psycopg itself is written. I'm not against that, but I wonder if this should be mentioned somewhere...

Mentioned this in the `NEWS`. Feedback on the wording is welcome.

If agreeable, a goal of mine to to do an eventual follow up of removing 2to3 from `lib/` as well. I believe this will be relatively straight forward and won't require the introduction of `six`, which I understand you don't want as a dependency. When it is time, I'll create a very small `compat.py` module, similar to the compat functions that have been created in `testutils`.

I am curious if you have a sense of how long you'd like to maintain Python 2.7 support.

jdufresne commented:
Thanks for the review and feedback!

Changes from the review

- Dropped `:$(PYTHONPATH)` in Makefile
- Fixed typo in `NEWS`
- Added more to `NEWS` about 2to3 changes
- Dropped unused `unicode` import
- Dropped accidental reintroduction of `skip_if_no_json_module`

All additional feedback is welcome.

---

Also, reverted back to installing for tests.

As you can see, this results in test failures due to conflicting import paths on Python 2. That is why I originally changed from `install` to `build`. If you have any thoughts on how to resolve this, let me know. But I'll try to find some time to discover a solution.

dvarrazzo commented:
From your work-in-progress it seems it was the `.egg-info` getting in the way. What about deleting it on `make clean` and running `make clean` before running the tests to run them against what's now in `site-packages`?

dvarrazzo commented:
> I am curious if you have a sense of how long you'd like to maintain Python 2.7 support?

Not really. I was planning for this holiday break to fix the bugs marked `2.7.4` and to release that. There is MR #576 which is still in progress (haven't heard from the OP yet about many of the changes I asked him) which is something that would reasonably end up in 2.8. But I don't see enough flesh for a new "major" release.

dvarrazzo commented:
Ha ha, [scripting windows is just hopeless](https://ci.appveyor.com/project/psycopg/psycopg2/build/2.x.262/job/9kctpvoyqpf00ga2#L244). I've stopped that build as it would have taken 8 hours to fail. And of course python doesn't have `make` so I guess removing the `.egg` is needed anyway.

jerickso commented:
Scripting windows, hopeless?  Maybe.  Hair pulling?  Definitely.

Adding a '/Q' to the command line should disable the confirmation, ie:
RD /S /Q
If one still wants to go down that route.

dvarrazzo commented:
> Scripting windows, hopeless? Maybe. Hair pulling? Definitely.

Utmost respect for who manages to do that and still be in good trichological conditions... :D

jdufresne commented:
Thanks for the windows scripting tips!

All tests are now passing after using `setup.py install`. Ready for any additional review.

dvarrazzo commented:
Hi, I don't understand very well how reviews work on github so I don't know if you have received that.

There is still one point open: not using the reverse-engineered build path but use a specific one. Are you interested in fixing it?

jdufresne commented:
> There is still one point open: not using the reverse-engineered build path but use a specific one. Are you interested in fixing it?

I'm happy to.

But, I'm not quite sure what kind of change you're asking for. Do you mind providing additional details or identifying the lines of the change that could be improved? Thanks.

dvarrazzo commented:
In the `Makefile` you have "fixed" the build dir by reverse-engineering the algorithm python uses, which is not part of an API:

```
PYTHON_PLATFORM := $(shell $(PYTHON) -c 'import sysconfig; print(sysconfig.get_platform())')
PYTHON_VERSION ?= $(shell $(PYTHON) -c 'import sys; print("%d.%d" % sys.version_info[:2])')
BUILD_DIR = $(shell pwd)/build/lib.$(PYTHON_PLATFORM)-$(PYTHON_VERSION)
```

the build dir is later used explicitly:

```
BUILD_OPT := --build-lib=$(BUILD_DIR)
BUILD_EXT_OPT := --build-lib=$(BUILD_DIR)
```

There are two possible cases:

1. the build dir is nowhere to be seen, it stays an implementation design of setuptools and we never access it. This is the case if the entire development/build/test chain doesn't need `BUILD_DIR` anymore, and all the commands using `BUILD_DIR` can be replaced by an equivalent not using it (e.g. is it possible to implement `make check` using `python setup.py develop`? Keep in mind that this build system is several years old, it predates `setuptools` etc. so it well be the case that things that we had to work around at time now are mainstream).

2. we still need `BUILD_DIR` explicitly somewhere: in this case it must be the `BUILD_DIR` we decide. All the pieces of the build chain already take the `BUILD_DIR` as an argument, so we don't need parroting what setuptools makes up.

So, if you want to fix the things along the lines of 2, it means reverting the Makefile as it was before this branch, dropping `PYTHON_PLATFORM`. If you want to fix the things along 1 it means dropping `BUILD_DIR` altogether.

The case 1 would be a simplification of the `Makefile` (maybe, but not necessarily, an improvement). 2. is the previous state of the affair: not pretty but it did its job. What proposed in this MR is more complex than before without fulfilling any specific purpose.

jdufresne commented:
Thank you for the details. That is very helpful. I'll give it a shot. I'm leaning towards option 1. at the moment, but I'll see how it goes.

jdufresne commented:
After investigation, it was clear that I misunderstood the use of the Makefile and the build directory definition. I have reverted the changes. Use `make && make check` now works correctly. In effect, this is taking approach 2 you've listed above.

I believe approach 1. will have additional ramifications that are outside the scope of this change. I still think there may be merit in approach 1, but would prefer to wait until after this lands before making such a large change.

This PR is ready for review. Thanks.

dvarrazzo commented:
It seems all fine, thank you very much. The `PYTHONPATH` in the makefile is back, but I can get rid of it no problem, and it's a thing independent from this MR.
